pkg/csi/service/server.go

```diff
@@ -0,0 +1,174 @@
+	// pending RPCs on the client side will get notified by connection
+	// errors.
+	Stop()
+
+	// GracefulStop stops the gRPC server gracefully. It stops the server
+	// from accepting new connections and RPCs and blocks until all the
+	// pending RPCs are finished.
+	GracefulStop()
+}
+
+// NewNonBlockingGRPCServer returns an instance of nonBlockingGRPCServer.
+func NewNonBlockingGRPCServer() NonBlockingGRPCServer {
+	return &nonBlockingGRPCServer{}
+}
+
+// nonBlockingGRPCServer implements the interface NonBlockingGRPCServer.
+type nonBlockingGRPCServer struct {
+	server *grpc.Server
+}
+
+func (s *nonBlockingGRPCServer) Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
+	log := logger.GetLoggerWithNoContext()
+	if err := s.serve(endpoint, ids, cs, ns); err != nil {
+		log.Errorf("failed to start grpc server. Err: %v", err)
+	}
+}
+
+func (s *nonBlockingGRPCServer) GracefulStop() {
+	log := logger.GetLoggerWithNoContext()
+	stopOnce.Do(func() {
+		if s.server != nil {
+			s.server.Stop()
+		}
+		log.Info("gracefully stopped")
+	})
+}
+
+func (s *nonBlockingGRPCServer) Stop() {
+	log := logger.GetLoggerWithNoContext()
+	stopOnce.Do(func() {
+		if s.server != nil {
+			s.server.Stop()
+		}
+		log.Info("stopped")
+	})
+}
+
+func (s *nonBlockingGRPCServer) serve(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) error {
+	log := logger.GetLoggerWithNoContext()
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		msg := fmt.Sprintf("failed to parse the endpoint %s. Err: %v", endpoint, err)
+		log.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	// CSI driver currently supports only unix path
+	if u.Scheme != "unix" {
+		msg := fmt.Sprintf("endpoint scheme %s not supported", u.Scheme)
+		log.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	addr := u.Path
+
+	// remove UNIX sock file if present.
+	if err := os.Remove(addr); err != nil && !os.IsNotExist(err) {
+		msg := fmt.Sprintf("failed to remove %s. Err: %v", addr, err)
+		log.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	listener, err := net.Listen(u.Scheme, addr)
+	if err != nil {
+		msg := fmt.Sprintf("failed to listen: %v", err)
+		log.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	server := grpc.NewServer()
+	s.server = server
+
+	// Register the CSI services.
+	// Always require the identity service.
+	if ids == nil {
+		return errors.New("identity service is required")
+	}
+	// Either a Controller or Node service should be supplied.
+	if cs == nil && ns == nil {
+		return errors.New("either a controller or node service is required")
+	}
+
+	// Always register the identity service.
+	csi.RegisterIdentityServer(s.server, ids)
+	log.Info("identity service registered")
+
+	// Determine which of the controller/node services to register
+	mode := os.Getenv(csitypes.EnvVarMode)
+	if strings.EqualFold(mode, "controller") {
+		mode = "controller"
+	} else if strings.EqualFold(mode, "node") {
+		mode = "node"
+	} else {
+		mode = ""
```

Comment:
Should `EnvVarMode` default to one of the two if not supplied?

Reply:
So if it's not supplied, we try to register both controller and node servers. So should we default to one of the two values?
Currently we have this variable supplied in all our manifests.

Reply:
I think we should fail if the mode is not supplied.

Reply:
I think it makes sense to make mode mandatory.

Reply:
Done. Changed the code to make it mandatory.